### PR TITLE
Pinning guide

### PR DESCRIPTION
**Pull request by SilentGardener**
Dear Alexandre, 

My suggestions for rephrasing the pinning guide, with the intent of making it more foolproof and easier to understand.

### CODE REVIEW

**Comment by alexandred** [comment]
Thank you for your contribution but there is much I take issue with this, in paritcular:

> If your APIC Pin Number is outside the range 0x5c-0x77, then VoodooI2C is not the correct driver for your touchpad. You are advised to seek support in the relevant forums or usergroups.

This is simply not true. *All* APIC pins are (at least in fairly well-tested theory) convertible to GPIO pins. For example, my APIC pin is 0x33 and I indeed have a valid corresponding GPIO pin.

Moreover, the _CRS method should not be completely overwritten as some people have critical device settings changed there. This is a mistake I made in the original GPIO pinning guide and have yet to upload the correction.

The separation into CRS-pinned and root-pinned is actually to make the guide more user friendly as I have had too many support requests asking "does it matter if my SBFB/SBFI/SBFG is in the root or in _CRS or in xyz".

With regards to the formula, I already have a different idea to make that section more user-friendly (although I already question whether someone should be fiddling with hackintosh if they can't even substitute values into a linear equation). I will be uploading a newer version of it soon.

**Comment by SilentGardener** [comment]
Hi Alexandre,

Your comments are valid, you are the expert and I am not, and my version has technical errors. 

However, I should explain that my proposal was not intended as a drop-in replacement, more as an example for a different structure of the guide and a different order of the steps. For instance, in your version, the attention of the reader has to switch from the hexadecimal APIC number, to the pin number in the to  the decimal GPIO number, and then back again to the  hexadecimal APIC number. Similarly with the section on renaming SBFI to SBFB after determining the APIC number and before calculating the GPIO pin number. Such context switches are costly for readers, it is best to avoid them.

Your version also does not make clear that entire guide's  purpose is just to end up with a correctly patched _CRS Return statement. Making that clear helps readers put the steps in context.

Finally, I took the trouble to try to understand the guide and I wrote down what I could make of it. The result is that my version has technical errors. There are two ways to look at those errors: they are my fault for being pig-ignorant in these matters, or, the guide is not clear enough for your target audience and leaves room for misunderstanding. 

I consider myself definitely to be your target audience: no expert on this matter itself, but reasonably technically experienced, for instance I customised Linux kernel drivers to drive LCD screens over GPIO and I2C to improve throughput, wrote firmware for USB HID devices, and I'm willing to learn sufficient ASL to patch the DSDT by hand. 

As I'm not very well versed on Mac OS kernel extensions, but I once made a living writing technical manuals, documentation is where I can help out.

So please consider my contribution in the spirit in which it is offered: as a suggestion for the structure of the guide, explanation of the purpose and organisation of the steps. You have the technical knowledge, I have the skills in how to communicate such knowledge to readers. I'm making an offer to help out with this guide, but your response gives me the feeling you do not wish to accept my offer. That is, of course, completely your own free choice and I take no offense.

Finally, I appreciate your work and all the effort you put into it.

Best,

SilentGardener

**Comment by alexandred** [comment]
Hi SilentGardner,

My intention was not to shut down your offer of help. My answers tend to be blunt and succinct due to time constraints on my own part. I realise that my wording can come across as dismissive and I do apologise for that - it is in no way meant to suggest that I am not open to PRs of any shape or form.

I should have been more clear that I am not dismissing your PR: I first wanted to update the guide with the full technical details necessary (as it exists in my head that is) as it is not yet complete (mainly accounting for some edge cases). The updating of the guide (and actual release of v2.0.0) were supposed to occur before Christmas but multiple setbacks arose with the driver itself and I have neglected to update the guide. Fortunately, I have now sorted out these issues and the driver is in a good-enough state for release. I am just waiting for some more people with the previously unsupported input devices to come online and test the changes.

As such, I'll be updating the guide with some changes which aim to streamline and remove confusion as well as correct some technical errors soon. I'll be more than happy for you to have a look over the new guide and submit a refreshed version of your changes - in that spirit I will reopen this PR.

Thanks,
Alex

**Comment by SilentGardener** [comment]
Hi Alex,

OK, I'll await the new draft!

Best,

SilentGardener

**Comment by xuhadixu** [comment]
because of this driver breaks laptops - be careful!

**Comment by xuhadixu** [comment]
> Thank you for your contribution but there is much I take issue with this, in paritcular:
> 
> > If your APIC Pin Number is outside the range 0x5c-0x77, then VoodooI2C is not the correct driver for your touchpad. You are advised to seek support in the relevant forums or usergroups.
> 
> This is simply not true. _All_ APIC pins are (at least in fairly well-tested theory) convertible to GPIO pins. For example, my APIC pin is 0x33 and I indeed have a valid corresponding GPIO pin.
> 
> Moreover, the _CRS method should not be completely overwritten as some people have critical device settings changed there. This is a mistake I made in the original GPIO pinning guide and have yet to upload the correction.
> 
> The separation into CRS-pinned and root-pinned is actually to make the guide more user friendly as I have had too many support requests asking "does it matter if my SBFB/SBFI/SBFG is in the root or in _CRS or in xyz".
> 
> With regards to the formula, I already have a different idea to make that section more user-friendly (although I already question whether someone should be fiddling with hackintosh if they can't even substitute values into a linear equation). I will be uploading a newer version of it soon.
because of this driver breaks laptops - be careful!

**Comment by xuhadixu** [comment]
> Hi SilentGardner,
> 
> My intention was not to shut down your offer of help. My answers tend to be blunt and succinct due to time constraints on my own part. I realise that my wording can come across as dismissive and I do apologise for that - it is in no way meant to suggest that I am not open to PRs of any shape or form.
> 
> I should have been more clear that I am not dismissing your PR: I first wanted to update the guide with the full technical details necessary (as it exists in my head that is) as it is not yet complete (mainly accounting for some edge cases). The updating of the guide (and actual release of v2.0.0) were supposed to occur before Christmas but multiple setbacks arose with the driver itself and I have neglected to update the guide. Fortunately, I have now sorted out these issues and the driver is in a good-enough state for release. I am just waiting for some more people with the previously unsupported input devices to come online and test the changes.
> 
> As such, I'll be updating the guide with some changes which aim to streamline and remove confusion as well as correct some technical errors soon. I'll be more than happy for you to have a look over the new guide and submit a refreshed version of your changes - in that spirit I will reopen this PR.
> 
> Thanks,
> Alex
/ Linux equivalent of FTE_i2c_write_cmd function
 IOReturn VoodooI2CFTETouchpadDriver::write_FTE_cmd(UInt16 reg, UInt16 cmd) {
     UInt16 buffer[] {
         reg,
         cmd
     };
     IOReturn retVal = kIOReturnSuccess;
     retVal = api->writeI2C(reinterpret_cast<UInt8*>(&buffer), sizeof(buffer));
     return retVal;
 }
 
 IOReturn VoodooI2CFTETouchpadDriver::message(UInt32 type, IOService* provider, void* argument)
 {
     switch (type)
     {
         case kKeyboardGetTouchStatus:
         {
 #if DEBUG
             IOLog("%s::getEnabledStatus = %s\n", getName(), ignoreall ? "false" : "true");
 #endif

File: Documentation/GPIO Pinning.md [hunk 1]
```diff
@@ -2,174 +2,163 @@
...
+	GpioInt (Level, ActiveLow, ExclusiveAndWake, PullDefault, 0x0000,
+	    "\\_SB.PCI0.GPI0", 0x00, ResourceConsumer, ,
+	    )
+	    {   // Pin list
+		0x0000
+	    }
+    })
 ```
 
-**you need `SBFG` and not `SBFI`**.
```
Comment (xuhadixu) [inline Documentation/GPIO Pinning.md]: / Linux equivalent of FTE_i2c_write_cmd function
 IOReturn VoodooI2CFTETouchpadDriver::write_FTE_cmd(UInt16 reg, UInt16 cmd) {
     UInt16 buffer[] {
         reg,
         cmd
     };
     IOReturn retVal = kIOReturnSuccess;
     retVal = api->writeI2C(reinterpret_cast<UInt8*>(&buffer), sizeof(buffer));
     return retVal;
 }
 
 IOReturn VoodooI2CFTETouchpadDriver::message(UInt32 type, IOService* provider, void* argument)
 {
     switch (type)
     {
         case kKeyboardGetTouchStatus:
         {
 #if DEBUG
             IOLog("%s::getEnabledStatus = %s\n", getName(), ignoreall ? "false" : "true");
 #endif

Comment (vcreatyv) [reply Documentation/GPIO Pinning.md]: kext is feik

**Comment by vcreatyv** [comment]
However I'm experiencing hidd to cause relatively high CPU usage with VoodooI2C 2.1 and 2.1.1 when the mouse cursor is being moved. With VoodooI2C 2.03 hidd CPU usage was around 5% for me. With latest releases it went up to 20%. Due to the CPU clock rate being raised, battery drain is caused.

**Comment by MarsArtis** [comment]
Hi there, the *.PNG screen in the documentation page are not viewable. They redirect to a 404 error page.

**Comment by alexandred** [comment]
> Hi there, the *.PNG screen in the documentation page are not viewable. They redirect to a 404 error page.

The documentation is intended to be viewed here: https://voodooi2c.github.io/ and not on the Github page.

**Pull request closed without merging**